polygeist/tools/cgeist/Lib/CGExpr.cc

```diff
@@ -1699,7 +1699,8 @@ ValueCategory MLIRScanner::VisitCastExpr(CastExpr *E) {
                                  Loc, Ty, Scalar, getConstantIndex(0)),
                              /*isReference*/ false);
 
-      if (ScalarTy.getShape().size() != MT.getShape().size()) {
+      if (ScalarTy.getShape().size() != MT.getShape().size() ||
+          ScalarTy.getElementType() != MT.getElementType()) {
```

Comment:
The problem was hidden by `memref2pointer(cast)->memref2pointer` folding canonicalization. 
memref.cast of different element types are not allowed.

Reply:
It would be nice to avoid using these 2 polygeist operations altogether, or at least limit the cases where we use them further. The memref dialect has the following operations that can be used to change the shape of a memref type:

- memref.reshape (converts between memrefs with same elem. type but different ranks, as long as the number of elems remains the same) 
- memref. expand_shape (produces a memref with higher rank) 
- memref.collapse_shape (produces a memref with smaller rank) 
- memref.realloc (can be used to expand the shape of a memref or to convert between memref<?xT> -> memref<1xT>

I don't think the memref dialect has an operation to change the element type of a memref type.  And such operations aren't safe w.r.t strict aliasing so they should be uncommon in the code (makes me think the compiler might have a bug somewhere that causes the memref types to have a different element.  Can you illustrate how that came about ? I want to be sure it is legal code and not a bug somewhere before we change this code.

Reply:
> I think you mean memref2pointer->pointer2memref

No, I meant `memref2pointer(cast)->memref2pointer `

I am adding the case where the element types are different, unrelated to the shape.

The case that I hit was:
```
memref.cast 
Source type:
'memref<?x!llvm.struct<(ptr<i8, 1>, ptr<i8, 1>, !sycl.id<[2], (!sycl.array<[2], (memref<2xi64, 4>)>)>, !sycl.id<[2], (!sycl.array<[2], (memref<2xi64, 4>)>)>, i64, i64, i64, i64)>>'
Target type:
 'memref<?x!llvm.struct<(ptr<i8, 4>, ptr<i8, 4>, !sycl.id<[2], (!sycl.array<[2], (memref<2xi64, 4>)>)>, !sycl.id<[2], (!sycl.array<[2], (memref<2xi64, 4>)>)>, i64, i64, i64, i64)>>'
 ```

Reply:
Added todo.